In [1]:
# import pandas
import pandas as pd

In [2]:
# read data
df = pd.read_csv('indian_food.csv')

# Завдання

## Про що дані?

Якщо звернути увагу на назву датасету та таблицю нижче, то не складно здогадатись, що ми працюємо з датасетом про різні індійські страви.

Про кожну страву ми знаємо її ім'я, і у більшості випадків знаємо які інгрідієнти потрібні для її приготування, час для підготовки та приготування, яка ця страва на смак, якою частиною трапези вона є та її походження.

## 1. Вивести `head() tail() sample() .shape, .columns, .index , .dtypes`. Пояснити зміст методів

Метод `df.head()` виводить `n` перших рядків таблиці. За замовчуванням `n` = 5. 

In [3]:
df.head()

name                                        ingredients  \
0      Balu shahi                    Maida flour, yogurt, oil, sugar   
1          Boondi                            Gram flour, ghee, sugar   
2  Gajar ka halwa       Carrots, milk, sugar, ghee, cashews, raisins   
3          Ghevar  Flour, ghee, kewra, milk, clarified butter, su...   
4     Gulab jamun  Milk powder, plain flour, baking powder, ghee,...   

         diet  prep_time  cook_time flavor_profile   course        state  \
0  vegetarian         45         25          sweet  dessert  West Bengal   
1  vegetarian         80         30          sweet  dessert    Rajasthan   
2  vegetarian         15         60          sweet  dessert       Punjab   
3  vegetarian         15         30          sweet  dessert    Rajasthan   
4  vegetarian         15         40          sweet  dessert  West Bengal   

  region  
0   East  
1   West  
2  North  
3   West  
4   East

Метод `df.tail()` виводить `n` останніх рядків таблиці. За замовчуванням `n` = 5. 

In [4]:
df.tail()

name                                        ingredients        diet  \
250  Til Pitha            Glutinous rice, black sesame seeds, gur  vegetarian   
251    Bebinca  Coconut milk, egg yolks, clarified butter, all...  vegetarian   
252     Shufta  Cottage cheese, dry dates, dried rose petals, ...  vegetarian   
253  Mawa Bati  Milk powder, dry fruits, arrowroot powder, all...  vegetarian   
254     Pinaca  Brown rice, fennel seeds, grated coconut, blac...  vegetarian   

     prep_time  cook_time flavor_profile   course            state      region  
250          5         30          sweet  dessert            Assam  North East  
251         20         60          sweet  dessert              Goa        West  
252         -1         -1          sweet  dessert  Jammu & Kashmir       North  
253         20         45          sweet  dessert   Madhya Pradesh     Central  
254         -1         -1          sweet  dessert              Goa        West

Метод `df.sample()` повертає рандомні елементи з таблиці. Найбільш базовий приклад використання - повертати `n` рандомних елементів. Параметр `n` за замовчуванням дорівнює 1.

In [5]:
df.sample()

name                     ingredients        diet  prep_time  \
7  Kaju katli  Cashews, ghee, cardamom, sugar  vegetarian         10   

   cook_time flavor_profile   course state region  
7         20          sweet  dessert    -1     -1

Атрибут `df.shape` повертає кортеж з двох значень: `(к-сть рядків, к-сть стовпців)`

In [6]:
print(f'DataFrame shape is: {df.shape}')

DataFrame shape is: (255, 9)


Атрибут `df.columns` повертає immutable sequence типу `pd.Index`. По ній можна ітеруватись і використовувати як індекс по стовпцям таблиці.

In [7]:
df.columns

Index(['name', 'ingredients', 'diet', 'prep_time', 'cook_time',
       'flavor_profile', 'course', 'state', 'region'],
      dtype='object')

Атрибут `df.index` повертає ітератор по лейблам рядків таблиці (в нашому випадку це числа від 0 до 254).

In [8]:
df.index

RangeIndex(start=0, stop=255, step=1)

Атрибут `df.dtypes` повертає об'єкт типу `pd.Series`. Його індекс - назви стовпців оригінальної таблиці, а значення - тип даних у відповідному стовпці.

In [9]:
df.dtypes

name              object
ingredients       object
diet              object
prep_time          int64
cook_time          int64
flavor_profile    object
course            object
state             object
region            object
dtype: object

## 2. Вибрати декілька колонок, рядків

Найпростіший спосіб ітерації по таблиці - використовуючи назву стовця, або список назв. Можемо також додати [python slicing notation](https://docs.python.org/3/reference/expressions.html?highlight=slicing#slicings) для того аби обрати потрібні рядки.

In [10]:
df[['name', 'region']][10:20]

name region
10        Laddu     -1
11        Lassi  North
12    Nankhatai     -1
13        Petha  North
14       Phirni   East
15        Rabri  North
16       Sheera   West
17      Singori  North
18  Sohan halwa  North
19  Sohan papdi   West

## 3. Вибрати декілька колонок, рядків використовуючи методи `loc`, `iloc`.

Метод `df.loc` дозволяє здійснювати індексацію по таблиці використовуюси лейбли рядків/стовпців. Ось так можемо отримати результат, аналогічний до того, що вище:

In [11]:
df.loc[10:19, ('name', 'region')]

name region
10        Laddu     -1
11        Lassi  North
12    Nankhatai     -1
13        Petha  North
14       Phirni   East
15        Rabri  North
16       Sheera   West
17      Singori  North
18  Sohan halwa  North
19  Sohan papdi   West

Окрім того, ми можемо використовувати різні умови, для більш "гнучкої" індексації. Ось так можна отримати результат вище, за виключенням стовпців, що дорівнюють `'-1'`.

In [12]:
df.loc[10:19, ('name', 'region')].loc[df['region'] != '-1']

name region
11        Lassi  North
13        Petha  North
14       Phirni   East
15        Rabri  North
16       Sheera   West
17      Singori  North
18  Sohan halwa  North
19  Sohan papdi   West

Метод `df.iloc` дозволяє здійснювати індексацію по таблиці використовуючи числовий індекс. Ось так можемо отримати результат, аналогічний до пункту `2`:

In [13]:
df.iloc[10:20, [0, 8]]

name region
10        Laddu     -1
11        Lassi  North
12    Nankhatai     -1
13        Petha  North
14       Phirni   East
15        Rabri  North
16       Sheera   West
17      Singori  North
18  Sohan halwa  North
19  Sohan papdi   West

## 4. Видалити непотрібні колонки (рядки)

Ми можемо видалити рядки або стовпці за допомогою методу `df.drop()`.

Щоб видалити рядки, в нашому випадку маємо вказати індекси рядків, які хочемо видалити.

In [14]:
# delete first 5 rows and make sure that now we have 250 columns instead of 255
print(f'Length of DataFrame after deleting 5 rows --> {len(df.drop(range(5)))}')

Length of DataFrame after deleting 5 rows --> 250


Щоб видалити стовпці, маємо вказати одну або декілька назв стовпців, та встановити `axis='columns'`

In [15]:
df.drop(['diet', 'state'], axis='columns').head()

name                                        ingredients  \
0      Balu shahi                    Maida flour, yogurt, oil, sugar   
1          Boondi                            Gram flour, ghee, sugar   
2  Gajar ka halwa       Carrots, milk, sugar, ghee, cashews, raisins   
3          Ghevar  Flour, ghee, kewra, milk, clarified butter, su...   
4     Gulab jamun  Milk powder, plain flour, baking powder, ghee,...   

   prep_time  cook_time flavor_profile   course region  
0         45         25          sweet  dessert   East  
1         80         30          sweet  dessert   West  
2         15         60          sweet  dessert  North  
3         15         30          sweet  dessert   West  
4         15         40          sweet  dessert   East

## 4. Створити нову колонку. В колонку записати суму значень двох колонок з таблиці.

Створимо новий стовпчик з сумою двох стовпців з числовими значеннями.

In [16]:
df['prep+cook_time'] = df['prep_time'] + df['cook_time']
df.head()

name                                        ingredients  \
0      Balu shahi                    Maida flour, yogurt, oil, sugar   
1          Boondi                            Gram flour, ghee, sugar   
2  Gajar ka halwa       Carrots, milk, sugar, ghee, cashews, raisins   
3          Ghevar  Flour, ghee, kewra, milk, clarified butter, su...   
4     Gulab jamun  Milk powder, plain flour, baking powder, ghee,...   

         diet  prep_time  cook_time flavor_profile   course        state  \
0  vegetarian         45         25          sweet  dessert  West Bengal   
1  vegetarian         80         30          sweet  dessert    Rajasthan   
2  vegetarian         15         60          sweet  dessert       Punjab   
3  vegetarian         15         30          sweet  dessert    Rajasthan   
4  vegetarian         15         40          sweet  dessert  West Bengal   

  region  prep+cook_time  
0   East              70  
1   West             110  
2  North              75  
3   West              45  
4   East              55

В цьому випадку стовчик додається в кінець таблиці. Ми можемо вставити стовчик в потрібне місце використовуючи метод [`df.insert()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.insert.html?highlight=insert#pandas.DataFrame.insert).

## 5. Відсортувати по збільшенню і по зменшенню

Ми можемо посортувати таблицю використовуючи метод `df.sort_values()`.

In [17]:
# sort values firstly by 'name' column, than by 'region'
df.sort_values(by=['name', 'region']).head()

name                                        ingredients  \
36          Adhirasam   Rice flour, jaggery, ghee, vegetable oil, elachi   
68          Aloo gobi  Cauliflower, potato, garam masala, turmeric, c...   
70         Aloo matar  Potato, peas, chillies, ginger, garam masala, ...   
71         Aloo methi      Potato, fenugreek leaves, chillies, salt, oil   
72  Aloo shimla mirch  Potato, shimla mirch, garam masala, amchur pow...   

          diet  prep_time  cook_time flavor_profile       course        state  \
36  vegetarian         10         50          sweet      dessert  West Bengal   
68  vegetarian         10         20          spicy  main course       Punjab   
70  vegetarian          5         40          spicy  main course       Punjab   
71  vegetarian         10         40         bitter  main course       Punjab   
72  vegetarian         10         40          spicy  main course       Punjab   

   region  prep+cook_time  
36   East              60  
68  North              30  
70  North              45  
71  North              50  
72  North              50

In [18]:
# sort values starting from the biggest 'cook time'
df.sort_values(by='cook_time', ascending=False).head()

name                                        ingredients  \
62     Shrikhand                     Curd, sugar, saffron, cardamom   
27       Malapua         Yoghurt, refined flour, ghee, fennel seeds   
75       Biryani  Chicken thighs, basmati rice, star anise, swee...   
114  Pindi chana   Fennel, tea bags, tomato, kasuri methi, cinnamon   
34      Rasgulla                            Chhena, sugar, cardamom   

               diet  prep_time  cook_time flavor_profile       course  \
62       vegetarian         10        720          sweet      dessert   
27       vegetarian         10        120          sweet      dessert   
75   non vegetarian         30        120          spicy  main course   
114      vegetarian        500        120          spicy  main course   
34       vegetarian         10         90          sweet      dessert   

           state region  prep+cook_time  
62   Maharashtra   West             730  
27         Bihar  North             130  
75     Telangana  South             150  
114       Punjab  North             620  
34   West Bengal   East             100

## 6. 3групувати по одній, двох, трьох колонках для різних значень (наприклад `max`). Агрегувати дані `agg()`

Агрегація, зазвичай, означає що ми беремо багато значень одного стовпця або рядка, виконуємо з ними певну операцію, і повертаємо одне агреговане значення.

В функцію `agg()` ми маємо передати функцію агрегації, або назву підтримуваної ф-ції та вказати над чим ми хочемо виконувати операцію: рядками чи стовпчиками. По замовчуванню ф-ція працює із стовпцями.

In [19]:
# get min, max and mean values of 3 numerical columns
df[['prep_time', 'cook_time', 'prep+cook_time']].agg(['min', 'max', 'mean'])

prep_time   cook_time  prep+cook_time
min    -1.000000   -1.000000       -2.000000
max   500.000000  720.000000      730.000000
mean   31.105882   34.529412       65.635294

## 7. Провести фільтрацію по значенню

Ми можемо фільтрувати таблицю по значенню використовуючи метод `df.filter()`. По замовчуванню метод фільтрує по стовпцям.

В якості критерія фільтрації ми можемо вказати 1 з 3 опцій:
- конкретні значення
- регулярний вираз
- умову типу `like`

In [20]:
# choose columns that starts with 'c'
df.filter(regex='^c').head()

cook_time   course
0         25  dessert
1         30  dessert
2         60  dessert
3         30  dessert
4         40  dessert

## 8. Об’єднати фільтрацію та групування

Метод `pd.groupby()` дозволяє нам спочатку згрупувати дані по спільним ознакам, а потім виконувати операції над кожною з груп.

In [21]:
# group dishes by state and then filter out all states that have less than 25 dishes
df.groupby('state').filter(lambda x: len(x) >= 25)['state'].unique()

array(['Punjab', 'Maharashtra', 'Gujarat'], dtype=object)

## 9. Отримати значення рядків з певним значенням параметру

In [22]:
# take all rows where ingredients contain sugar and the dish is not a dessert
df.loc[(df['ingredients'].str.contains('sugar')) & (df['course'] != 'dessert')].head()

name                                        ingredients  \
134  Keerai masiyal  Urad dal, curry leaves, sugar, mustard seeds, ...   
155           Puttu            Brown rice flour, sugar, grated coconut   
164            Upma   Chana dal, urad dal, ginger, curry leaves, sugar   
174          Chevdo       Citric acid, fry, raisins, sugar, chana daal   
190     Keri no ras                                       Mango, sugar   

           diet  prep_time  cook_time flavor_profile       course       state  \
134  vegetarian         -1         -1          spicy  main course  Tamil Nadu   
155  vegetarian        495         40             -1  main course      Kerala   
164  vegetarian         10         20          spicy        snack          -1   
174  vegetarian         20         30          spicy        snack     Gujarat   
190  vegetarian         10         10           sour  main course     Gujarat   

    region  prep+cook_time  
134  South              -2  
155  South             535  
164     -1              30  
174   West              50  
190   West              20

## 10. Написати довільний запит `query`

Метод `pd.query()` дозволяє нам отримувати рядки з таблиці використовуючи дещо спрощений синтакс запиту. Наприклад, цей синтакс дозволяє нам використовувати значення зовнішних змінних прямо у запиті.

In [23]:
# write query that shows dishes that require no more than max_prep_time and max_cook_time from us.
max_prep_time = 20
max_cook_time = 40
df.query('prep_time <= @max_prep_time and cook_time <= @max_cook_time').head()

name                                        ingredients        diet  \
3       Ghevar  Flour, ghee, kewra, milk, clarified butter, su...  vegetarian   
4  Gulab jamun  Milk powder, plain flour, baking powder, ghee,...  vegetarian   
7   Kaju katli                     Cashews, ghee, cardamom, sugar  vegetarian   
8     Kalakand                        Milk, cottage cheese, sugar  vegetarian   
9        Kheer                    Milk, rice, sugar, dried fruits  vegetarian   

   prep_time  cook_time flavor_profile   course        state region  \
3         15         30          sweet  dessert    Rajasthan   West   
4         15         40          sweet  dessert  West Bengal   East   
7         10         20          sweet  dessert           -1     -1   
8         20         30          sweet  dessert  West Bengal   East   
9         10         40          sweet  dessert           -1     -1   

   prep+cook_time  
3              45  
4              55  
7              30  
8              50  
9              50

## 11. Кожний метод бібліотеки pandas має багато параметрів. Продемонструвати вплив параметрів на результат. (3 приклади)

### 11.1 Параметр `inplace`

В першому прикладі я продемонструю суть параметра `inplace`. Я покажу його використання на методі `df.drop()`

In [24]:
copy_df = df.copy()  # copy dataframe
copy_df.drop('name', axis='columns').head(1)

ingredients        diet  prep_time  cook_time  \
0  Maida flour, yogurt, oil, sugar  vegetarian         45         25   

  flavor_profile   course        state region  prep+cook_time  
0          sweet  dessert  West Bengal   East              70

In [25]:
copy_df.head(1)

name                      ingredients        diet  prep_time  \
0  Balu shahi  Maida flour, yogurt, oil, sugar  vegetarian         45   

   cook_time flavor_profile   course        state region  prep+cook_time  
0         25          sweet  dessert  West Bengal   East              70

Як можемо бачити, ми викликали метод `df.drop()` і відобразили таблицю без стовпця `name`. Потім ви відобразили таблицю ще раз, і стовпчик повернувся назад. Це відбувається тому що за замовчуванням метод `df.drop()` (і ще багато інших) повертає нову таблицю.

Додамо параметр `inplace=True`.

In [26]:
copy_df.drop('name', axis='columns', inplace=True)

In [27]:
copy_df.head(1)

ingredients        diet  prep_time  cook_time  \
0  Maida flour, yogurt, oil, sugar  vegetarian         45         25   

  flavor_profile   course        state region  prep+cook_time  
0          sweet  dessert  West Bengal   East              70

Тепер ми змінили таблицю "на місці", видаливши з неї вказаний стовпчик.

### 11. 2 Метод `df.filter()` параметр `items`

У пункті `7` я вже демонстрував як працює фільтр з параметром `regex`. Розглянемо як працює фільтр з параметром `items`. Нагадаю, що в цьому параметрі ми маємо вказати назви конкретних стовпців, по яким хочемо відфільтрувати.

In [28]:
df.filter(items=['state', 'region']).head()

state region
0  West Bengal   East
1    Rajasthan   West
2       Punjab  North
3    Rajasthan   West
4  West Bengal   East

### 11. 3 Метод `df.filter()` параметр `like`

У пункті `7` я вже демонстрував як працює фільтр з параметром `regex`. Розглянемо як працює фільтр з параметром `like`. Нагадаю, що в цьому параметрі ми маємо вказати таке значення, щоб воно проходило перевірку типу `like in label == True`.

In [29]:
df.filter(like='me').head()

name  prep_time  cook_time  prep+cook_time
0      Balu shahi         45         25              70
1          Boondi         80         30             110
2  Gajar ka halwa         15         60              75
3          Ghevar         15         30              45
4     Gulab jamun         15         40              55

Як можемо бачити, ми отримали всі стовпці, у яких зустрічається фрагмент `ме`.